src/pnpq/devices/waveplate_thorlabs_kb10crm.py

```diff
@@ -343,7 +343,7 @@ def custom_home(self, degree: float | int) -> None:
     def custom_rotate(self, degree: float | int) -> None:
         """Rotation with customized home!"""
 
-        if not self.relative_home:
+        if self.relative_home is None:
```

Comment:
Unfortunately, this isn't going to work. I should have noticed this myself, but I'm still learning the rules of Python variable definition myself, and I forgot about "truthy" and "falsy" values as well.

I think that the best way to explain it is with some simple example code:

```python
#!/usr/bin/env python

class MyClass():
    set_variable = "class_set_variable"
    unset_variable: str

    def __init__(self):
        self.set_variable = "instance_set_variable"

    def uncalled_function(self):
        self.unset_variable = "unset_variable_value"

    def print_values(self):
        print("class-level set variable: " + MyClass.set_variable)
        print("instance-level set variable: " + self.set_variable)

        # print("class-level unset variable: " + MyClass.unset_variable)
        # print("instance-level unset variable" + self.unset_variable)
        # if not self.unset_variable:
        #    print("not unset_variable")
        if self.unset_variable is None:
            print("unset_variable is None")

if __name__ == "__main__":
    my_obj = MyClass()
    my_obj.print_values()
```

You can play around with the commented-out code a little to see the different errors that you get.

This becomes even more confusing if you change `unset_variable: str` to `unset_variable: str = None` ! That will create a class-level variable called `unset_variable`, and then `self.unset_variable` will work as you expect.

However, this won't pass a type checker. `None` is not type `str`.

And then Python [dataclasses](https://docs.python.org/3/library/dataclasses.html) work according to different rules again. I won't explore that here.

So, what is the solution?

We need to ensure that `self.relative_home` always has a value.

For now, the easiest way to do this is to add a line to `__init__`:

```
self.relative_home = 0.0
```

and then completely remove lines 346 to 353.

Reply:
Thank you for providing such a helpful description.